python/cudf/cudf/core/dataframe.py

```diff
@@ -638,34 +639,25 @@ def __dir__(self):
         return list(o)
 
     def __setattr__(self, key, col):
-
-        # if an attribute already exists, set it.
         try:
+            # Preexisting attributes may be set. We cannot rely on checking the
+            # `_PROTECTED_KEYS` because we must also allow for settable
+            # properties, and we must call object.__getattribute__ to bypass
+            # the `__getitem__` behavior inherited from `GetAttrGetItemMixin`.
             object.__getattribute__(self, key)
-            object.__setattr__(self, key, col)
-            return
+            super().__setattr__(key, col)
         except AttributeError:
-            pass
-
-        # if a column already exists, set it.
-        if key not in self._internal_names:
-            try:
-                self[key]  # __getitem__ to verify key exists
-                self[key] = col
-                return
-            except KeyError:
-                pass
-
-        object.__setattr__(self, key, col)
-
-    def __getattr__(self, key):
-        if key in self._internal_names:
-            return object.__getattribute__(self, key)
-        else:
-            if key in self:
-                return self[key]
-
-        raise AttributeError("'DataFrame' object has no attribute %r" % key)
+            if key not in self._PROTECTED_KEYS:
+                try:
+                    # If a column already exists, set it.
+                    self[key]  # __getitem__ to verify key exists
+                    self[key] = col
```

Comment:
Not part of this PR but would it be better to use `self._data[key] = col` here? Looks like a special case we can bypass `__getitem__` overhead.

Reply:
No, unfortunately not. This code is a bit more general this way because it also works in cases where `self` does not have an `_data` attribute. In particular, it's being used by `Rolling`.

Reply:
Oh whoops, I missed that this was in `__setitem__`, not the shared `__getitem__`. Yes, I think that would work, let me give it a shot.

Reply:
I've modified both the above lines to take faster paths. I've tried this and my tests pass locally, but I'm not 100% sure that the set is exactly equivalent. I'm reasonably confident that replacing `self[key]` with `self._data._select_by_label_grouped` is completely equivalent here based on the possible values of `key` in this context, but `DataFrame.__setitem__` even for scalar keys seems a bit more complex and I'm a bit concerned that there might be some kinds of inputs for which this won't work. Maybe @shwina would know for sure?

Reply:
Discussed with @shwina offline. This is _probably_ safe, but there are enough edge cases that it's hard to be entirely sure. I'm going to change this back for now but keep it in mind in case this ever shows up in profiles.